WireDomain/Sources/WireDomain/NotificationService/NotificationService.swift

```diff
@@ -0,0 +1,189 @@
+
+        onGoingTask = Task {
+            do {
+                let notificationUserInfo = request.content.userInfo
+
+                let notification = try NotificationPayload(
+                    userInfo: notificationUserInfo
+                )
+
+                notificationSession = try await createNotificationSession(
+                    userID: notification.userID
+                )
+
+                try await notificationSession?.processPushNotification(
+                    eventID: notification.eventID
+                )
+
+            } catch {
+                logError(error)
+                finishWithEmptyNotification()
+            }
+        }
+    }
+
+    override func serviceExtensionTimeWillExpire() {
+        logger.warn("legacy service extension will expire")
+        finishWithEmptyNotification()
+    }
+
+    private func createNotificationSession(
+        userID: UUID
+    ) async throws -> NotificationSession {
+        let userLocalStore: UserLocalStoreProtocol = Injector.resolve()
+        let selfUserInfo = await userLocalStore.selfUserInfo()
+        let environment: BackendEnvironmentProvider = Injector.resolve()
+
+        let cookieStorage = ZMPersistentCookieStorage(
+            forServerName: environment.backendURL.host!,
+            userIdentifier: userID,
+            useCache: false
+        )
+
+        let isAuthenticated = cookieStorage.isAuthenticated
+
+        guard isAuthenticated else {
+            throw Failure.notAuthenticated
+        }
+
+        guard let selfClientID = selfUserInfo.clientId else {
+            throw Failure.missingSelfClientID
+        }
+
+        let updateEventsRepository = UpdateEventsRepository(
+            userID: userID,
+            selfClientID: selfClientID,
+            // these were already initialized, resolving them
+            updateEventsAPI: Injector.resolve(),
+            pushChannel: Injector.resolve(),
+            updateEventDecryptor: Injector.resolve(),
+            updateEventsLocalStore: Injector.resolve()
+        )
+
+        return NotificationSession(
+            updateEventsRepository: updateEventsRepository
+        ) { [weak self] notificationContent in
+            self?.finishWithNotification(content: notificationContent)
+        }
+    }
+
+    private func finishWithNotification(content: UNNotificationContent) {
+        // TODO: [WPB-11175]
+    }
+
+    private func finishWithEmptyNotification() {
+        logger.info("finishing without showing notification")
+        let emptyNotification = UNNotificationContent()
+
+        // With the "filtering" entitlement, we can tell iOS to not display a user notification by
+        // passing empty content to the content handler.
+        // See https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_usernotifications_filtering
+        contentHandler?(emptyNotification)
+        terminate()
+    }
+
+    private func logError(_ error: any Error) {
```

Comment:
Perhaps we can add `errorDescription` to all the errors and just log them inline.